src/Compilers/CSharp/Test/Semantic/Semantics/PropertyFieldKeywordTests.cs

```diff
@@ -1873,15 +1930,89 @@ public double P
-            }
+        [Fact]
+        public void SpeculativeSemanticModel_FieldKeywordInRegularAccessor_ReplaceBlock_MethodBodySemanticModel_BindOriginalFirst()
+        {
+            var comp = CreateCompilation(@"
+class C
+{
+    public int P
+    {
+        get
+        {
+            return 0;
+        }
+    }
+}
+");
+            var accessorBindingData = new SourcePropertySymbolBase.AccessorBindingData();
+            comp.TestOnlyCompilationData = accessorBindingData;
+            Assert.Empty(comp.GetTypeByMetadataName("C").GetMembers().OfType<FieldSymbol>());
+            comp.VerifyDiagnostics();
+
+            var tree = comp.SyntaxTrees[0];
+            var token = tree.GetRoot().DescendantTokens().Single(t => t.IsKind(SyntaxKind.NumericLiteralToken));
+
+            var model = comp.GetSemanticModel(tree);
+            var accessor = SyntaxFactory.ParseSyntaxTree(@"
+class C
+{
+    public int P
+    {
+        get
+        {
+            return field;
+        }
+    }
+}").GetRoot().DescendantNodes().OfType<AccessorDeclarationSyntax>().Single();
+            model.TryGetSpeculativeSemanticModelForMethodBody(token.SpanStart, accessor, out var speculativeModel);
+            var fieldKeywordSymbolInfo = speculativeModel.GetSymbolInfo(accessor.DescendantNodes().OfType<IdentifierNameSyntax>().Single());
+            Assert.Null(fieldKeywordSymbolInfo.Symbol);
+            Assert.Empty(comp.GetTypeByMetadataName("C").GetFieldsToEmit());
+            Assert.Equal(0, accessorBindingData.NumberOfPerformedAccessorBinding);
+        }
+
+        [Fact]
+        public void SpeculativeSemanticModel_FieldKeywordInRegularAccessor_ReplaceBlock_MethodBodySemanticModel_BindSpeculatedFirst()
+        {
+            var comp = CreateCompilation(@"
+class C
+{
+    public int P
+    {
+        get
+        {
+            return 0;
+        }
+    }
+}
+");
+            var accessorBindingData = new SourcePropertySymbolBase.AccessorBindingData();
+            comp.TestOnlyCompilationData = accessorBindingData;
+
+            var tree = comp.SyntaxTrees[0];
+            var token = tree.GetRoot().DescendantTokens().Single(t => t.IsKind(SyntaxKind.NumericLiteralToken));
+
+            var model = comp.GetSemanticModel(tree);
+            var accessor = SyntaxFactory.ParseSyntaxTree(@"
+class C
+{
+    public int P
+    {
+        get
+        {
+            return field;
+        }
+    }
+}").GetRoot().DescendantNodes().OfType<AccessorDeclarationSyntax>().Single();
+            model.TryGetSpeculativeSemanticModelForMethodBody(token.SpanStart, accessor, out var speculativeModel);
+            var fieldKeywordSymbolInfo = speculativeModel.GetSymbolInfo(accessor.DescendantNodes().OfType<IdentifierNameSyntax>().Single());
+            Assert.Null(fieldKeywordSymbolInfo.Symbol);
+            Assert.Empty(comp.GetTypeByMetadataName("C").GetFieldsToEmit());
+            Assert.Equal(0, accessorBindingData.NumberOfPerformedAccessorBinding);
```

Comment:
We should add some tests for the following sets of options:

1. `GetSpeculativeSymbolInfo, with both `SpeculativeBindingOption`s being tested (`IdentifierName` is a `TypeSyntax`, so it should do something, even if it will return no info because `field` isn't used as a type). 
2. `GetSpeculativeTypeInfo`, similarly with both `SpeculativeBindingOption`s.
3. `GetSpeculativeAliasInfo`, similarly with both `SpeculativeBindingOption`s.
4. `TryGetSpeculativeSemanticModel` that takes a `TypeSyntax`, again with both `SpeculativeBindingOption`s.
5. `TryGetSpeculativeSemanticModel` that takes an `EqualsValueClauseSyntax`, used as a parameter initializer (in a local function inside the property, for example). I know we haven't done all the work for binding nested contexts yet, it's ok if this doesn't quite return the correct results. Just leave a prototype comment.

For these cases, we want to test on the axes of:
1. `field` existing as a keyword vs `field` not existing as a keyword in the original compilation
2. What option we're passing to `SpeculativeBindingOption`.
3. Original bound first vs speculative bound first (you've already done this for some of these).
4. Nullable completely disabled vs enabled (this affects whether we bind things completely before answering semantic model questions).

Some of these axes can likely be tested by a `Theory` test (such as nullable enabled vs disabled), and some will need to be tested as separate tests.

Reply:
@333fred @AlekseyTs I updated the existing tests for `SpeculativeBindingOption.BindAsTypeOrNamespace`. Can you review the behavior? Meanwhile, I'll work on adding tests for `GetSpeculativeTypeInfo`, `GetSpeculativeAliasInfo`, `EqualsValueClauseSyntax`.

I think `GetSpeculativeSymbolInfo` and `TryGetSpeculativeSemanticModel` that takes `TypeSyntax` are covered now, but let me know if anything is missing (other than testing with nullable enabled and disabled)

Reply:
Added more tests. The remaining is `TryGetSpeculativeSemanticModel` that takes `EqualsValueClauseSyntax` and nullable.

Reply:
> > `TryGetSpeculativeSemanticModel` that takes an `EqualsValueClauseSyntax`, used as a parameter initializer (in a local function inside the property, for example).
> 
> We probably also have something like this for attributes that could be applied to local functions.

Is this what you mean?

```csharp
using System;

public class C : Attribute
{
    public int Test { get; set; }

    public int P
    {
        get
        {
            return local(); // or field + local();

            [C(Test = 0)] // speculate with field here.
            static int local() => 0;
        }
    }
}
```

Note that I think it doesn't have to do with `EqualsValueClauseSyntax`. Attributes use `NameEqualsSyntax` instead.

Reply:
> Is this what you mean?

I mean speculate with an ```AttributeSyntax``` that uses `field` in one of its expressions.  
```
public bool TryGetSpeculativeSemanticModel(int position, AttributeSyntax attribute, out SemanticModel speculativeModel)
```

Reply:
Basically I think we need to test all speculation related APIs at position within an accessor.

Reply:
@AlekseyTs Nice! This revealed a missing `SpeculativeFieldKeywordBinder`. I also fixed a broken assert and added another assert in `AttributeSemanticModel` in e481a2dfe497173639654535692f7c8b628b303c. Can you take another look please? Thanks!